gbpservice/nfp/tests/unit/nfp/configurator/test_firewall.py

```diff
@@ -0,0 +1,887 @@
+        super(ConfiguratorRpcManagerTestCase, self).__init__(*args, **kwargs)
+        self.fo = FakeObjects()
+
+    @mock.patch(__name__ + '.FakeObjects.conf')
+    @mock.patch(__name__ + '.FakeObjects.sc')
+    def _get_ConfiguratorRpcManager_object(self,sc, conf):
+        cm = cfgr.ConfiguratorModule(sc)
+        demuxer = demuxer_lib.ConfiguratorDemuxer()
+        rpc_mgr = cfgr.ConfiguratorRpcManager(sc, cm, conf, demuxer)
+        return sc, conf, rpc_mgr
+
+    def _get_GenericConfigRpcManager_object(self, conf, sc):
+        agent = gc.GenericConfigRpcManager(sc, conf)
+        return agent, sc
+
+    @mock.patch(__name__ + '.FakeObjects.nqueue')
+    @mock.patch(__name__ + '.FakeObjects.drivers')
+    def _get_GenericConfigEventHandler_object(self, sc, rpcmgr, drivers, nqueue):
+        agent = gc.GenericConfigEventHandler(sc, drivers, rpcmgr, nqueue)
+        return agent
+
+    def _get_FWaasRpcManager_object(self, conf, sc):
+        agent = fw.FWaasRpcManager(sc, conf)
+        return agent, sc
+
+    def _test_generic_event_creation(self, operation, method, bulk=False):
+        sc, conf, rpc_mgr = self._get_ConfiguratorRpcManager_object()
+        agent, sc = self._get_GenericConfigRpcManager_object(conf, sc)
+        
+        if bulk:
+            request_data = self.fo.fake_request_data_generic_bulk()
+            request_data1 = self.fo.fake_request_data_generic_bulk()
+        else:
+            request_data = self.fo.fake_request_data_generic_single()
+            request_data1 = self.fo.fake_request_data_generic_single()
+
+        if (method == 'CONFIGURE_ROUTES' or method == 'CLEAR_ROUTES'):
+            request_data1['config'][0]['resource'] = 'routes'
+            request_data['config'][0]['resource'] = 'routes'
+
+        with mock.patch.object(
+                    sc, 'new_event', return_value='foo') as mock_sc_event, \
+            mock.patch.object(sc, 'post_event') as mock_sc_rpc_event, \
+            mock.patch.object(rpc_mgr,
+                              '_get_service_agent_instance',
+                              return_value=agent):
+            if operation == 'create':
+                rpc_mgr.create_network_device_config(
+                                    self.fo.context, request_data)
+            elif operation == 'delete':
+                rpc_mgr.delete_network_device_config(
+                                    self.fo.context, request_data)
+
+            context = request_data1['config'][0]['kwargs']['context']
+            context.update({'notification_data': {}})
+            context.update({'resource': request_data1['config'][0]['resource']})
+            del request_data1['config'][0]['kwargs']['context']
+            kwargs = request_data1['config'][0]['kwargs']
+            if bulk:
+                sa_info_list = self.fo.fake_sa_info_list()
+                if operation == 'delete':
+                    sa_info_list[0]['method'] = 'clear_interfaces'
```

Comment:
We need to write two test cases for demuxer methods to test the construction of sa_info_list.